viewing_party/party.py

```diff
@@ -1,23 +1,153 @@
+
+
+def get_most_watched_genre(user_data): 
+    fav_genre = []
+    counter = 0
+    top_genre = fav_genre
+
+    for i in user_data["watched"]:
+        fav_genre.append(i["genre"])
+    
+    for genre in fav_genre:
+        curr_freq = fav_genre.count(genre)
+        if(curr_freq > counter):
+            counter = curr_freq
+            top_genre = genre
+            return top_genre
+        elif fav_genre == []:
+            return None
 
 
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
+def get_unique_watched(user_data):
+    list_of_movies = []
+    all_friend_movies = []
+
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            if movie not in all_friend_movies:
+                all_friend_movies.append(movie)
+
+    for movie in user_data["watched"]:
+        if movie not in all_friend_movies:
+            list_of_movies.append(movie)
+    return list_of_movies    
+    
+def get_friends_unique_watched(user_data):
+    user_movies = []
+    unique_friend_movies = []
+
+    for movie in user_data["watched"]:
+        if movie not in user_movies:
+            user_movies.append(movie)
+    
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            if movie not in user_movies and movie not in unique_friend_movies:
+                unique_friend_movies.append(movie)
+    return unique_friend_movies
 
-        
 # -----------------------------------------
 # ------------- WAVE 4 --------------------
 # -----------------------------------------
+def get_available_recs(user_data):
+    friends_movies = get_friends_unique_watched(user_data)
+    user_subscriptions = []
+    list_of_recommendations = []
+
+    for user in user_data["subscriptions"]:
+        user_subscriptions.append(user)
+    for item in friends_movies:
+        if item["host"] in user_subscriptions:
+            list_of_recommendations.append(item)
+    return list_of_recommendations
 
 # -----------------------------------------
 # ------------- WAVE 5 --------------------
 # -----------------------------------------
 
+def get_new_rec_by_genre(user_data):
+    fave_user_genre = get_most_watched_genre(user_data)
+    recommended_list = []
+    unique_watched = get_friends_unique_watched(user_data)
+    
+
+    if not user_data["watched"]:
+        return recommended_list
+    for movie in unique_watched:
+        if movie["genre"] in fave_user_genre:
+            recommended_list.append(movie)
+    return recommended_list
+
+def get_rec_from_favorites(user_data):
+    list_of_movie_recs = []
+    friend_watch_list = []
+    
+    for movie in user_data["friends"]:
+        for item in movie["watched"]:
+            friend_watch_list.append(item)
+
+    for movie in user_data["favorites"]:
+        if movie not in friend_watch_list:
+            list_of_movie_recs.append(movie)
+    return list_of_movie_recs 

```

Comment:
Arguably, this is a missed chance to reuse `get_friends_unique_watched`. I say "arguably", because it's possible that the user has movies in `user_data["favorites"]` that are not in `user_data["watched"]`, and that would justify the lack of reuse.

I notice that we do not make any statement about that in the problem, but it is likely reasonable to assume that a movie is not in `user_data["favorites"]` unless it is also in `user_data["watched"]`. And I think we do intend that you will be able to make that assumption, but it's not a big miss on your part.